pbsmetrics/pbsmetrics.go

```diff
@@ -0,0 +1,350 @@
+// testing routines to ensure that no metrics are written anywhere.
+
+func NewBlankMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName) *Metrics {
+	newMetrics := &Metrics{
+		metricsRegistry:     registry,
+		RequestMeter:        blankMeter(0),
+		AppRequestMeter:     blankMeter(0),
+		NoCookieMeter:       blankMeter(0),
+		SafariRequestMeter:  blankMeter(0),
+		SafariNoCookieMeter: blankMeter(0),
+		ErrorMeter:          blankMeter(0),
+		InvalidMeter:        blankMeter(0),
+		RequestTimer:        blankTimer(0),
+		CookieSyncMeter:     blankMeter(0),
+		ORTBRequestMeter:    blankMeter(0),
+
+		AdapterMetrics: make(map[openrtb_ext.BidderName]*AdapterMetrics, len(exchanges)),
+
+		exchanges: exchanges,
+	}
+	for _, a := range exchanges {
+		newMetrics.AdapterMetrics[a] = makeBlankAdapterMetrics(registry, a)
+	}
+
+	return newMetrics
+}
+
+// Create a new Metrics object with needed metrics defined. In time we may develop to the point
+// where Metrics contains all the metrics we might want to record, and then we build the actual
+// metrics object to contain only the metrics we are interested in. This would allow for debug
+// mode metrics. The code would allways try to record the metrics, but effectively noop if we are
+// using a blank meter/timer.
+
+func NewMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName) *Metrics {
+	newMetrics := NewBlankMetrics(registry, exchanges)
+	newMetrics.RequestMeter = metrics.GetOrRegisterMeter("requests", registry)
+	newMetrics.SafariRequestMeter = metrics.GetOrRegisterMeter("safari_requests", registry)
+	newMetrics.ErrorMeter = metrics.GetOrRegisterMeter("error_requests", registry)
+	newMetrics.NoCookieMeter = metrics.GetOrRegisterMeter("no_cookie_requests", registry)
+	newMetrics.AppRequestMeter = metrics.GetOrRegisterMeter("app_requests", registry)
+	newMetrics.SafariNoCookieMeter = metrics.GetOrRegisterMeter("safari_no_cookie_requests", registry)
+	newMetrics.RequestTimer = metrics.GetOrRegisterTimer("request_time", registry)
+	newMetrics.ORTBRequestMeter = metrics.GetOrRegisterMeter("ortb_requests", registry)
+
+	for _, a := range exchanges {
+		registerAdapterMetrics(registry, "adapter", string(a), newMetrics.AdapterMetrics[a])
+	}
+	return newMetrics
+}
+
+// Part of setting up blank metrics, the adapter metrics.
+func makeBlankAdapterMetrics(registry metrics.Registry, exchanges openrtb_ext.BidderName) *AdapterMetrics {
+	newAdapter := &AdapterMetrics{
+		NoCookieMeter:     blankMeter(0),
+		ErrorMeter:        blankMeter(0),
+		NoBidMeter:        blankMeter(0),
+		TimeoutMeter:      blankMeter(0),
+		RequestMeter:      blankMeter(0),
+		RequestTimer:      blankTimer(0),
+		PriceHistogram:    blankHistogram(0),
+		BidsReceivedMeter: blankMeter(0),
+	}
+	return newAdapter
+}
+
+func registerAdapterMetrics(registry metrics.Registry, adapterOrAccount string, exchange string, am *AdapterMetrics) {
```

Comment:
I know this is copy/pasted, but... this is only called in one place, and the second argument is `adapter`, so... the logic can be streamlined.

Reply:
I decided to leave this as its own function in case we ever decide to go back to account level metrics. If we are fairly sure we will not have a use case for that in the future, I can certainly inline this bit.

Reply:
No need. I don't have any real insight into the future here.